runtime/evaluate.cc

```diff
@@ -414,42 +414,93 @@ class evaluator {
+          // anything (it might already have been finished by other workers). To work around this, we initialize the
+          // context on the first iteration we run.
+          if (context.size() == 0) {
+            if (loop->closure) {
+              // The body is a closure, so we know exactly which symbols we need to copy to the new local context.
+              context.reserve(loop->context.size());
+              context.config = loop->context.config;
+
+              // Assume that this let_stmt is a closure for this loop. We'll evaluate the values using the parent
+              // context, but assign them to our local context.
+              for (const std::pair<var, expr>& i : loop->closure->lets) {
+                if (i.first == loop->sym) {
+                  // The loop variable is part of the closure, because it is defined outside the closure and used inside
+                  // it. However, we are going to overwrite it below.
+                  continue;
+                }
+                auto src = as_variable(i.second);
+                assert(src);
+                context.set(i.first, loop->context.lookup(*src));
               }
-              context.set(i.first, evaluate(i.second, *parent_context));
+            } else {
+              // We don't have a closure, just copy the whole context.
+              context = loop->context;
             }
-          } else {
-            // We don't have a closure, just copy the whole context.
-            context = *parent_context;
           }
-          context.set(op->sym, i * step + min);
+
+          context.set(loop->sym, i * loop->step + loop->min);
           // Evaluate the parallel loop body with our copy of the context.
-          index_t result_i = evaluate(op->body, context);
+          index_t result_i = evaluate(loop->body, context);
           if (result_i != 0) {
             index_t zero = 0;
-            result.compare_exchange_strong(zero, result_i);
+            loop->result.compare_exchange_strong(zero, result_i);
           }
-        },
-        op->max_workers);
-    return result;
+        });
+        // If we get here, there's no more work to start.
+        if (context.size() != 0) {
+          // We ran some tasks, so we know that the context is still in scope. Cancel any remaining tasks.
+          context.config->thread_pool->cancel(loop.get());
+        }
+      };
+      int workers = std::min<int>(op->max_workers, std::min<std::size_t>(pool->thread_count() + 1, n));
+      if (workers > 1) {
+        pool->enqueue(workers - 1, worker, loop.get());
+      }
+      // Running the worker here guarantees forward progress on the loop even if no threads in the thread pool are
+      // available.
+      pool->run(worker, loop.get());
+      // While the loop still isn't done, work on other tasks.
+      pool->wait_for([&]() { return loop->done(); });
+
+      return loop->result;
```

Comment:
Maybe it's possible to add another variant of the thread_pool::parallel_for which takes a worker function to avoid some of the code duplication?

Reply:
I thought about it, but I think we should just work towards eliminating parallel_for from the thread pool entirely. It's only used by benchmarks/tests now anyways. We can move such benchmarks/tests to construct slinky programs (like test/evaluate.cc)